### PR TITLE
Mark SupportAlertFragment and AlertFragment as Public to fix crash is…

### PR DESCRIPTION
**Pull request by lesliesam**
Hi,
This changelist is for fixing a crash issue on Android devices. For detail please see, https://github.com/facebook/react-native/issues/7080

In short, the crash occurred when a RN app with a Alert dialog shown at front switches back after it is destroyed at background due to lack of phone memory.

My fix is to let those Alert fragment classes accessible from android.support.v4.app.Fragment instantiate() function, so that it won't crash. And since other UI will be reloaded whatever, mListener is set to null to avoid any callback happen.

### CODE REVIEW

**Comment by lesliesam** [comment]
OK, I have signed that cla.

File: ReactAndroid/src/main/java/com/facebook/react/modules/dialog/AlertFragment.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 /**
  * A fragment used to display the dialog.
  */
-/* package */ class AlertFragment extends DialogFragment implements DialogInterface.OnClickListener {
+public class AlertFragment extends DialogFragment implements DialogInterface.OnClickListener {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/dialog/AlertFragment.java]: Why make this class public?

File: ReactAndroid/src/main/java/com/facebook/react/modules/dialog/AlertFragment.java [hunk 2]
```diff
@@ -32,6 +32,10 @@
 
   private final @Nullable DialogModule.AlertFragmentListener mListener;
 
+  public AlertFragment() {
+    mListener = null;
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/dialog/AlertFragment.java]: You don't need this line, it's initialized to null by default.

File: ReactAndroid/src/main/java/com/facebook/react/modules/dialog/SupportAlertFragment.java [hunk 3]
```diff
@@ -21,10 +21,14 @@
  * {@link AlertFragment} for apps that use the Support FragmentActivity and FragmentManager
  * for legacy reasons.
  */
-/* package */ class SupportAlertFragment extends DialogFragment implements DialogInterface.OnClickListener {
+public class SupportAlertFragment extends DialogFragment implements DialogInterface.OnClickListener {
 
   private final @Nullable DialogModule.AlertFragmentListener mListener;
 
+  public SupportAlertFragment() {
+    mListener = null;
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/dialog/SupportAlertFragment.java]: You don't need this line, it's initialized to null by default.

**Comment by mkonicek** [comment]
Thanks for the detailed repro steps and for the fix @lesliesam!

1. Show an Alert dialog with whatever content. 
2. Switches RN app into background
3. Open as much apps as possible on your phone and exhaust all your phone memory . (A way to know if the app is destroyed is looking into the Android Studio debug if the app's process no longer available in the list.)
4. Switch back to the RN app so that onCreate() will be called to restore the dialog. At that time, this crash will happen.

With this fix, what happens in step 4? Is the dialog shown with the correct contents? Can you interact with it? Can you close it?

**Comment by lesliesam** [comment]
With this fix, the dialog will shown with the correct content and it can be closed as normal.

**Comment by satya164** [comment]
@lesliesam If the whole UI is recreated, why keep the dialog only? Also if no callbacks attached to the dialog will fire when you click on a button, it's not the expected behaviour, and can be quite confusing.

**Comment by lesliesam** [comment]
This change is about to fix a crash bug. It might be a little confuse, but it is better than a oops crashed message. What do you think is the better solution for get rid of the crash?

**Comment by lesliesam** [comment]
I have a work around to avoid show the dialog. Call super.onCreate(null); instead of super.onCreate(savedInstanceState); in the ReactActivity.onCreate function. I guess that only works for me or a few apps as others might have logic rely on the savedInstanceState.

What's your though over this?

**Comment by mkonicek** [comment]
Let's merge this to avoid the crash but agree the behavior can be a bit confusing. PRs with a more general fix (maybe don't show the dialog) welcome.

@facebook-github-bot shipit

**Comment by mkonicek** [comment]
Removing the import failed label because the code had a legit compile error the first time we tried to shipit.

@facebook-github-bot shipit

**Comment by facebook-github-bot-3** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1991559147736480/int_phab) to review.

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

**Comment by mkonicek** [comment]
@lesliesam The code doesn't compile, see https://circleci.com/gh/facebook/react-native/6314

**Comment by facebook-github-bot-2** [comment]
@lesliesam updated the pull request.

**Comment by lesliesam** [comment]
Compile error should be fixed.

**Comment by facebook-github-bot-0** [comment]
@mkonicek would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-2** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1991559147736480/int_phab) to review.

**Pull request closed without merging**